component/models/sdjwt/common/verification.go

```diff
@@ -0,0 +1,392 @@
+	}
+
+	return nil
+}
+
+func contains(values []string, val string) bool {
+	for _, v := range values {
+		if v == val {
+			return true
+		}
+	}
+
+	return false
+}
+
+// VerifyJWT checks that the JWT is valid using nbf, iat, and exp claims (if provided in the JWT).
+func VerifyJWT(signedJWT *afgjwt.JSONWebToken, leeway time.Duration) error {
+	var claims jwt.Claims
+
+	d, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Result:           &claims,
+		TagName:          "json",
+		Squash:           true,
+		WeaklyTypedInput: true,
+		DecodeHook:       utils.JSONNumberToJwtNumericDate(),
+	})
+	if err != nil {
+		return fmt.Errorf("mapstruct verifyJWT. error: %w", err)
+	}
+
+	if err = d.Decode(signedJWT.Payload); err != nil {
+		return fmt.Errorf("mapstruct verifyJWT decode. error: %w", err)
+	}
+
+	// Validate checks claims in a token against expected values.
+	// It is validated using the expected.Time, or time.Now if not provided
+	expected := jwt.Expected{}
+
+	err = claims.ValidateWithLeeway(expected, leeway)
+	if err != nil {
+		return fmt.Errorf("invalid JWT time values: %w", err)
+	}
+
+	return nil
+}
+
+// VerifyTyp checks JWT header parameters for the SD-JWT component.
+func VerifyTyp(joseHeaders jose.Headers, expectedTyp string) error {
+	typ, ok := joseHeaders.Type()
+	if !ok {
+		return fmt.Errorf("missing typ")
+	}
+
+	if typ != expectedTyp {
+		return fmt.Errorf("unexpected typ \"%s\"", typ)
+	}
+
+	return nil
+}
+
+// VerifyDisclosuresInSDJWT checks for disclosure inclusion in SD-JWT.
+func VerifyDisclosuresInSDJWT(
+	disclosures []string,
+	signedJWT *afgjwt.JSONWebToken,
+) error {
+	claims := utils.CopyMap(signedJWT.Payload)
+
+	cryptoHash, err := GetCryptoHashFromClaims(claims)
+	if err != nil {
+		return err
+	}
+
+	parsedDisclosureClaims, err := getDisclosureClaims(disclosures, cryptoHash)
+	if err != nil {
+		return err
+	}
+
+	recData := &recursiveData{
+		disclosures:          parsedDisclosureClaims,
+		cleanupDigestsClaims: false,
+	}
+
+	_, err = discloseClaimValue(claims, recData)
+	if err != nil {
+		return err
+	}
+
+	// If the digest cannot be found in the SD-JWT payload, the Verifier MUST reject the Presentation.
+	for _, disclosure := range parsedDisclosureClaims {
+		if !disclosure.IsValueParsed {
+			return fmt.Errorf("disclosure digest '%s' not found in SD-JWT disclosure digests", disclosure.Digest)
+		}
+	}
+
+	return nil
+}
+
+func setDisclosureClaimValue(recData *recursiveData, disclosureClaim *DisclosureClaim) error {
+	if disclosureClaim.IsValueParsed {
+		return nil
+	}
+
+	newValue, err := discloseClaimValue(disclosureClaim.Value, recData)
+	if err != nil {
+		return err
+	}
+
+	disclosureClaim.Value = newValue
+	disclosureClaim.IsValueParsed = true
+
+	return nil
+}
+
+// discloseClaimValue returns new value of claim, resolving dependencies on other disclosures.
+func discloseClaimValue(claim interface{}, recData *recursiveData) (interface{}, error) { // nolint:funlen,gocyclo
```

Comment:
`discloseClaimValue` is the root function of this patch.
It responsible for:
- resolving recursive digests in disclosures.
- resolving digests in incoming claims. 

also, it does a basic checks, e.g.
- claim key duplications on the same nesting level
- digests duplications
- disclosure elements length (2 elems for array digest, 3 elems for sd digest).